### ISSUE TITLE
[$500] PDF receipt preview loads each time visiting 1:1 DM, report and transaction thread

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#37852]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.48-0
**Reproducible in staging?:** y
**Reproducible in production?:** n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
1. Go to staging.new.expensify.com
2. Go offline.
3. Go to FAB > Request money > Scan.
4. Select a PDF receipt.
5. Select any participant and create the Scan request.
6. in 1:1 DM, click on the scan preview.
7. In IOU report, click on the scan preview.
8. Click on the header subtitle to return to report and 1:1 DM.

## Expected Result:
The PDF receipt preview will not load each time going back and forth between 1:1 DM, IOU report and transaction thread.

## Actual Result:
The PDF receipt preview loads each time going back and forth between 1:1 DM, IOU report and transaction thread. Sometimes it shows blue "loading page" briefly when the receipt is rendering.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/43996225/885644dc-0d8b-43ab-86a3-b288dfe8d92d

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016df3671b30e63c80</li>
        <li>Upwork Job ID: 1806060107499225786</li>
        <li>Last Price Increase: 2024-07-24</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

**Comment by kavimuru** [comment]
@anmurali FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by deetergp** [comment]
Is this our culprit PR? https://github.com/Expensify/App/pull/35255 (cc @eh2077 @s77rt @luacmartins)

**Comment by youssef-lr** [comment]
We have just added support for PDF receipts, so I'm not sure this would qualify as a deploy blocker because we don't support PDF receipts in production right now.

**Comment by deetergp** [comment]
Thanks @youssef-lr, that sounds reasonable. I'll remove the label and demote this back down to Daily.

**Comment by s77rt** [comment]
@eh2077 Can you please take a quick look on this?

**Comment by eh2077** [comment]
@s77rt Sure. I'm looking into it.

**Comment by eh2077** [comment]
It seems not easy to avoid reloading PDF when going back and forth between 1:1 DM, IOU report and transaction thread. This is because the browser by default can't cache the `canvas` of the PDFThumbnail but it's able to cache `image` by default.

While checking this issue, I also found another flaw that's easier to avoid - PDF thumbnail reloads when hovering on it


https://github.com/Expensify/App/assets/117511920/df575660-eba5-4a00-9dc1-08f9e0fda1dd

**Comment by eh2077** [comment]
I think the reloading issue when hovering is a legit regression. I can make a quick PR to fix it.

But the reloading issue when re-visiting different pages seems a tricky one. Can this be a followup improvement instead of a regression? 

@s77rt @luacmartins What're your opinions?

**Comment by s77rt** [comment]
@eh2077 Can you please raise a PR fixing the hover issue? I thought we fixed that here https://github.com/Expensify/App/pull/35255#discussion_r1469048195 Is this another case that we missed?

**Comment by luacmartins** [comment]
Yea, let's fix the hover issue and then investigate the reload one as a separate issue.

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/35255
- The offending PR has been commented on: N/A (Bug fixed by the author)
- A discussion in #expensify-bugs has been started: Not needed.
- Determine if we should create a regression test for this bug: No

**Comment by s77rt** [comment]
@luacmartins / @deetergp We still have the bug when switching reports; let's reopen this since we only fixed the hover case.

**Comment by anmurali** [comment]
I think we should wait till the bug is fixed on switching reports as well before we pay this. Do you agree @s77rt ?

**Comment by s77rt** [comment]
Yes. In fact we should reopen this. Can you or @deetergp add `Help Wanted` label, remove `Awaiting Payment`  and update the title

**Comment by deetergp** [comment]
Should be all set @s77rt.

**Comment by deetergp** [comment]
Going to make this a weekly till we have some proposals.

**Comment by deetergp** [comment]
Still waiting on proposals

**Comment by jsdev2547** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
PDF receipt preview loads each time visting DM, report and transaction thread.

### What is the root cause of that problem?
The react-pdf document is loading url using the file prop. According to react-pdf, 

```
({ url: 'http://example.com/document.pdf' }) === ({ url: 'http://example.com/document.pdf' })
false
```
When the expected would have been true.

react-pdf takes the file as referentially unequal each time the file prop is set to previewSourceURL and causes re-render.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L22

### What changes do you think we should make in order to solve the problem?
Since only the thumbnail preview of the first page is needed andimage receipt previews are loading correctly. I suggest to convert the first page of the pdf receipt to image for previewing purposes and render the first page as image rather than rendering the first page as pdf. The pdf approach would be difficult given the contraints posed by react-pdf.

The approach uses helper method to get the first page of the pdf receipt and render to canvas element and then get the DataURL (img in base64 format). The base64 img is stored in localstorage and uses fileName as key to the base64 img.

The PDFThumbnail component would need fileName of the pdf receipt for mapping in localStorage. Using URl.createObjectURL returns different url each time which causes re-render so need to map using fileName.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L15

```
    const isImageLoadedRef = useRef(false);
    const [imageUrl, setImageUrl] = useState<string>(null);
    const data = localStorage.getItem(fileName);
    if(data && !isImageLoadedRef.current)
    {
        isImageLoadedRef.current = true;
        setImageUrl(data);
    }
    useEffect(() => {
        if (data) {
            return;
        }
        const canvas = document.createElement("canvas");
        pdfjs.getDocument(url).promise.then((pdf) => {
            for (let i = 0; i < 1; i++) {
                pdf.getPage(i + 1).then((page) => {
                    const viewport = page.getViewport({ scale: 5 });
                    const context = canvas.getContext("2d");
                    canvas.height = viewport.height;
                    canvas.width = viewport.width;
                    var renderContext = {
                        canvasContext: context,
                        viewport: viewport
                    };

                    page.render(renderContext).promise.then((e) => {
                        const data = canvas.toDataURL();
                        localStorage.setItem(fileName, data);
                        isImageLoadedRef.current = true;
                        setImageUrl(data);
                        canvas.remove();
                    });
                })
            }
        });
    }, [data])
```

In the PDFThumbnail component instead of loading react-pdf component load ReceiptImage component and pass the base64 imageURL.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L40

```
    {imageUrl &&
        <ReceiptImage
            isThumbnail={false}
            source={imageUrl}
            // AuthToken is required when retrieving the image from the server
            // but we don't need it to load the blob:// or file:// image when starting a money request / split bill
            // So if we have a thumbnail, it means we're retrieving the image from the server
            isAuthTokenRequired={isAuthTokenRequired}
        />
    }
```

### What alternative solutions did you explore? (Optional)

**Comment by jsdev2547** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
PDF receipt preview loads each time visting DM, report and transaction thread.

### What is the root cause of that problem?
The react-pdf document is loading url using the file prop. According to react-pdf, 

```
({ url: 'http://example.com/document.pdf' }) === ({ url: 'http://example.com/document.pdf' })
false
```
When the expected would have been true.

react-pdf takes the file as referentially unequal each time the file prop is set to previewSourceURL and causes re-render.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L22

### What changes do you think we should make in order to solve the problem?
Since only the thumbnail preview of the first page is needed and image receipt previews are loading correctly. I suggest to convert the first page of the pdf receipt to image for previewing purposes and render the first page as image rather than rendering the first page as pdf. The pdf approach would be difficult given the contraints posed by react-pdf.

The approach uses helper method to get the first page of the pdf receipt and render to canvas element and then get the DataURL (img in base64 format). The base64 img is stored in localstorage and uses fileName as key to the base64 img.

The PDFThumbnail component would need fileName of the pdf receipt for mapping in localStorage. Using URl.createObjectURL returns different url each time which causes re-render so need to map using fileName.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L15

```
    const isImageLoadedRef = useRef(false);
    const [imageUrl, setImageUrl] = useState<string>(null);
    const data = localStorage.getItem(fileName);
    if(data && !isImageLoadedRef.current)
    {
        isImageLoadedRef.current = true;
        setImageUrl(data);
    }
    useEffect(() => {
        if (data) {
            return;
        }
        const canvas = document.createElement("canvas");
        pdfjs.getDocument(url).promise.then((pdf) => {
            for (let i = 0; i < 1; i++) {
                pdf.getPage(i + 1).then((page) => {
                    const viewport = page.getViewport({ scale: 5 });
                    const context = canvas.getContext("2d");
                    canvas.height = viewport.height;
                    canvas.width = viewport.width;
                    var renderContext = {
                        canvasContext: context,
                        viewport: viewport
                    };

                    page.render(renderContext).promise.then((e) => {
                        const data = canvas.toDataURL();
                        localStorage.setItem(fileName, data);
                        isImageLoadedRef.current = true;
                        setImageUrl(data);
                        canvas.remove();
                    });
                })
            }
        });
    }, [data])
```

In the PDFThumbnail component instead of loading react-pdf component load ReceiptImage component and pass the base64 imageURL.
https://github.com/Expensify/App/blob/b9bd07fcbf13ad517a0b42b5a4dc4abafefb91f7/src/components/PDFThumbnail/index.tsx#L40

```
    {imageUrl &&
        <ReceiptImage
            isThumbnail={false}
            source={imageUrl}
            // AuthToken is required when retrieving the image from the server
            // but we don't need it to load the blob:// or file:// image when starting a money request / split bill
            // So if we have a thumbnail, it means we're retrieving the image from the server
            isAuthTokenRequired={isAuthTokenRequired}
        />
    }
```

### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@jsdev2547 Thanks for the proposal. Regarding the RCA

> The react-pdf document is loading url using the file prop. According to react-pdf,
>```
> ({ url: 'http://example.com/document.pdf' }) === ({ url: 'http://example.com/document.pdf' })
> false
> ```

Can you link to the relevant code? If the bug is in react-pdf then it should be fixed upstream

**Comment by jsdev2547** [comment]
@s77rt I downgraded to react-pdf version 6.2.2 and the pdf thumbnail is still reloading when navigation from DM to report and transaction thread. I also checked if the route navigation was causing re-render when navigating from DM to report and transaction thread, but when image receipt is used even though the route navigates in reportPreview and MoneyRequestAction no reloading issue is observed, but the reloading issue is present when using pdf thumbnails. So the issue is definitely the third party lib react-pdf. But the important thing is not to try and solve the bug in third party lib but to correct approach of displaying pdf thumbnails.

Route navigation
https://github.com/Expensify/App/blob/8f90c2567f020ed924a96593244727aace9a7d72/src/components/ReportActionItem/ReportPreview.tsx#L257

https://github.com/Expensify/App/blob/8f90c2567f020ed924a96593244727aace9a7d72/src/components/ReportActionItem/MoneyRequestAction.tsx#L96

More so what I have observed is the correct approach to solving the issue of pdf thumbnails is to generate thumbnail of pdf by rendering the pdf page to canvas and then converting the canvas to image using toDataURL() for preview.

pdfjs in the provided example of of pdf thumbnail has suggested to convert the canvas to image and then display the image, rather than displaying the pdf page directly.

https://github.com/mozilla/pdf.js/blob/master/web/pdf_thumbnail_view.js#L216

**Comment by s77rt** [comment]
@jsdev2547 Let's fix the root cause. Rendering pdf then converting to image is more work than just rendering the pdf.

**Comment by jsdev2547** [comment]
@s77rt The pdf would not be rendered, i.e. the react-pdf document component would not be used in pdf thumbnail. The poc is quite straight forward, get page 1 of pdf, render on canvas, get the image from canvas. The root cause would be to not to render pdf thumbnails as pdf but to render pdf thumbnails as image. I was trying to actually use pdfjs to render pdf thumbnail but then came across pdf_thumbnail example which suggests to convert pdf to thumbnail rather use pdf directly. I have poc video of pdf image thumbnails in pdf receipts and how the functionality appears, if needed for demo purposes.

**Comment by deetergp** [comment]
@jsdev2547 I'm pretty sure our back end creates a thumbnail of these uploaded PDFs automagically. If you replace the `.pdf` extension of the uploaded PDF with `.jpg` there you should have a JPG rendering of the fist page of the PDF. And if you add `.320.jpg` or `.1024.jpg` on to that you'll have 320px & 1024px thumbnails. 

You may not have access to actually view these but I guarantee you they are there:
https://staging.expensify.com/receipts/w_edee87a4d07915c51e6b58ed256da890177f2a3c.pdf
https://staging.expensify.com/receipts/w_edee87a4d07915c51e6b58ed256da890177f2a3c.jpg
https://staging.expensify.com/receipts/w_edee87a4d07915c51e6b58ed256da890177f2a3c.jpg.320.jpg
https://staging.expensify.com/receipts/w_edee87a4d07915c51e6b58ed256da890177f2a3c.jpg.1024.jpg

**Comment by s77rt** [comment]
@jsdev2547 
> get page 1 of pdf, render on canvas, get the image from canvas

This is more work than what we have currently in place where we just render the pdf on canvas.

**Comment by s77rt** [comment]
@deetergp When offline we don't have access to those urls. We render the pdf instead

**Comment by jsdev2547** [comment]
@deetergp Interesting, then the conversion step is already done only the thumbnail needs displaying.

**Comment by jsdev2547** [comment]
@s77rt the issue is regarding the thumbnail reloading not offline, so @deetergp might have point on using the urls provided by server. I need to look into and get back.

**Comment by jsdev2547** [comment]
@deetergp @s77rt The urls become available after the scan request is complete and the urls are available offline and online. The issue is regarding reloading of the pdf which happens either online or offline while the user is waiting for the scan request to complete. Once the scan request is complete the backend sends the thumbnail of the first page of the pdf and the pdfthumbnail component is no longer used instead the pdf image thumbnail is rendered using thumbnail component. The isPDFThumbnail variable is also set to false for pdf receipts and shouldUseThumbnailImage is set to true.

https://github.com/Expensify/App/blob/d4bc1b08daa1890829461ab464ef4894ce343d06/src/components/ReceiptImage.tsx#L113

The backend already has the correct approach of sending thumbnails of first page of pdf receipts after scan is complete, while the user is waiting for the scan to complete(online or offline), the front end should also correct the approach and generate image thumbnail of pdf. The thumbnail images generated on the front end would resolve the reloading issue and be available (online or offline) while the user is waiting for scan request to complete. Once the scan request is complete, the thumbnail would be served from the backend.

So what needs to be decided is to whether to push for and continue the use of pdfs in thumbnails while the user is waiting for the scan to complete, or whether the front end should be consistent with the backend and generate thumbnails for pdf while the scan is in progress. I vote for generating thumbnail of first page pdf on the front end because the effort is inlined with the backend and not hard to implement. I already having functional solution which solves the issue (I can provide video and branch for testing). Trying to get the pdf to directly render in thumbnail(also spending time to find bug in react-pdf) seems to be unnecessary effort when the server would eventually return thumbnails. I would have agreed more with the second approach if the backend was not sending thumbnails.

**Comment by s77rt** [comment]
@jsdev2547 Thanks for looking into this. Generating an image has to pass via the pdf rendering phase. This is not giving us any advantage and it's only resulting in more work. We should just fix the pdf rendering phase to not re-render if the source is not changed.

**Comment by jsdev2547** [comment]
> Generating an image has to pass via the pdf rendering phase. 

@s77rt The current implementation is going to load the pdf document anyways on first render, the entire image generation can be done in onLoadSuccess event and is not difficult to implement. Please take moment to review the implementation and test the provided branch.

https://github.com/jsdev2547/App/tree/generatePdfThumbnail

Also added poc video,

[poc.webm](https://github.com/Expensify/App/assets/161593779/8f8fc200-efdd-4e0f-8fed-2944d245e5ad)

> We should just fix the pdf rendering phase to not re-render if the source is not changed.

Typically the right approach would be to correct the useEffect to not re-render if the source has not changed, but in the given issue, the useEffect is in third party lib and also I have concerns the approach of rendering pdf directly for thumbnails is incorrect. react-pdf is using pdf.js and in the pdf.js example I mentioned is also converting first page to image thumbnail and also the backend is generating image thumbnail. Also I find little advantage in trying to display the pdf thumbnail when after the scan is complete the backend would be sending image thumbnails anyways. so the correct approach should be to simply display image thumbnails (online and offline) while the user is waiting for the scan to complete.

@deetergp perhaps internal engineer can be requested to get second opinion on whether pdf should be used for thumbnails while the user is waiting for scan to complete (online & offline) or whether front end should generate image thumbnails and after the scan is complete display the thumbnails sent by backend.

**Comment by jsdev2547** [comment]
@s77rt The example from the pdf.js for thumbnail previews suggests to use image thumbnails and not render pdf directly. I am curious what @deetergp and the internal engineer opinion would be on generating image thumbnails (online and offline).

**Comment by s77rt** [comment]
Not overdue. Still looking for proposals

**Comment by s77rt** [comment]
Same

**Comment by deetergp** [comment]
Making weekly while await additional proposals.

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
Same ^

**Comment by deetergp** [comment]
MOTS

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
Same ^

**Comment by deetergp** [comment]
Still no proposals, but that may be because there wasn't an External label on this one. Added it now.

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by deetergp** [comment]
@anmurali I think we need to adjust the bounty on this one.

**Comment by s77rt** [comment]
Still looking for proposals...

**Comment by deetergp** [comment]
Since we are still seeking proposals, I'm going to bump this down to weekly.

**Comment by anmurali** [comment]
Increased the bounty to $500 on the job.

**Comment by tsa321** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

In Offline mode, PDF thumbnail of receipt is always re-rendered when revisiting the report

### What is the root cause of that problem?

When showing PDF thumbnail of the receipt in ReportActionItemImage:

https://github.com/Expensify/App/blob/fa738c641e7d2b289850da7b4be11369c33b96d6/src/components/ReportActionItem/ReportActionItemImage.tsx#L111-L112

the PDFThumbnail:

https://github.com/Expensify/App/blob/fa738c641e7d2b289850da7b4be11369c33b96d6/src/components/PDFThumbnail/index.tsx#L15

does't cache the rendered image. So the pdf file will be rendered each time the pdfthumbnail mounted.

### What changes do you think we should make in order to solve the problem?

We could use `canvasRef` property of Thumbail of the `PDFThumnail`:

https://github.com/Expensify/App/blob/fa738c641e7d2b289850da7b4be11369c33b96d6/src/components/PDFThumbnail/index.tsx#L42

and use the canvasElement to generate URL blob image of the canvas.

The rough code could be, in PDFThumbnail we add canvas ref and  use `canvasRef` and `onRenderSuccess` property of thumbnail: 

```diff

+ const canvas = useRef();
...
<Thumbnail 
      pageIndex={0} 
+      canvasRef={canvas} 
+      onRenderSuccess={() => onRenderSuccess(canvas.current)}
}} />

```

Then in ReportActionItemImage we add:

```javascript

const cachedPDFReceiptThumbnail = useRef(ReceiptUtils.getPDFThumbnailCache(transaction.transactionID));
...
...
const cachePDFThumbnail = useCallback((canvasElement) => {
    if (!!canvasElement) {
        canvasElement.toBlob((blob) => {
            const href = URL.createObjectURL(blob);
            cachedPDFReceiptThumbnail.current = href;
            ReceiptUtils.cachePDFReceiptThumbnail(transaction.transactionID, href);
        }, 'image/png');
    }
}, [transaction]);


....
else if (isLocalFile && filename && Str.isPDF(filename) && typeof attachmentModalSource === 'string') {
    propsObj = {
        isPDFThumbnail: true,
        source: attachmentModalSource,
        onPDFRenderSuccess: cachePDFThumbnail,
        cachedPDFThumbnail: cachedPDFReceiptThumbnail.current
    };
}
```
The ReceiptUtils.cachePDFReceiptThumbnail could be simple :

```javaascript
function cachePDFReceiptThumbnail(transactionID: string, pdfThumbnailCache: string) {
    pdfThumbnailImageCaches[transactionID] = pdfThumbnailCache;  
}

function getPDFThumbnailCache(transactionID: string) {
    return pdfThumbnailImageCaches[transactionID];
}
```

**Comment by tsa321** [comment]
## Proposal

[updated](https://github.com/Expensify/App/issues/37852#issuecomment-2241075167)

**Comment by s77rt** [comment]
The bug is no longer reproducible. The PDF preview loads only when going forth (opening new reports), when going back the pdf preview is kept rendered as expected.


https://github.com/user-attachments/assets/eac28b11-f487-4167-a7ee-50fb313c303e



@deetergp We can close this one.

**Comment by s77rt** [comment]
@tsa321 Going back to the IOU report via the "From" link does not cause the preview to reload, but doing so from the LHN reloads the preview. Can you explain why is that happening?


https://github.com/user-attachments/assets/53f31fd6-a5ad-4f9d-8950-bdddbf994fc2

**Comment by tsa321** [comment]
@s77rt, please wait. I will try to take a look.

**Comment by s77rt** [comment]
Waiting on proposals...

**Comment by tsa321** [comment]
@s77rt The PDF also loads immediately when the user presses the browser's back button. When the user presses the link to the parent report in the report header, it functions similarly to going back (much like pressing the browser's back button). Perhaps the stack navigation/browser preserves the content of the previous screen/route, allowing the PDF to load immediately.

**Comment by s77rt** [comment]
@tsa321 I don't think we really have a bug then. Opening a new report renders a **new** PDF Previewer and the loader is expected since the pdf is loaded at library level (unlike images where they are loaded at the browser level where caching is [implemented](https://github.com/chromium/chromium/blob/162a46d71be9646d2836b10804977e9314fb0701/components/image_fetcher/core/cache/image_cache.cc)). This can be a feature request but if we implement a caching mechanism it should be at `react-pdf` level and we should cache the pdf content and not its thumbnail.

(Same note applies for mobile if the bug is reproducible there)

**Comment by tsa321** [comment]
@s77rt Actually, we can cache all pages of the PDF as images. Is this the expected result? Caching the PDF means converting it into images, correct? Also, the server returns the PDF thumbnail as images too. Why should we do this at the `react-pdf` level? We can do it in the app.

**Comment by s77rt** [comment]
@tsa321 Caching is to save the pdf data (bytes) for a given uri. If we use a known uri we'd use the cached data instead of fetching that uri contents again.

> Why should we do this at the `react-pdf` level? 

It's a feature for `react-pdf` and not related to the app.

For this case the feature is a low priority since the uri that the PDF Previewer will load always points to a local file

**Comment by tsa321** [comment]
@s77rt, the PDF file is already stored as a blob URI, meaning it's stored in memory as bytes.

I believe we may have different perspectives on PDF caching:
My understanding is that PDF rendering involves drawing various elements such as text, lines, shapes, rectangles, paths, vectors, and images specified in the PDF file onto a canvas. This process is managed by react-pdf or whichever PDF engine library is utilized (indicated by the loading spinner during rendering).

On the other hand, caching the PDF means storing the rendered result as an image, so that it doesn't need to be redrawn.

However, I agree that this issue is of very low priority. I'm okay with closing this issue. @deetergp, what are your thoughts on this issue? Thank you.

**Comment by s77rt** [comment]
@deetergp Let's close this based on https://github.com/Expensify/App/issues/37852#issuecomment-2241601923

**Comment by deetergp** [comment]
Sounds like a plan @s77rt 👍
